stdlib/multiprocessing/__init__.pyi

```diff
@@ -10,12 +10,10 @@ from multiprocessing.context import (
 from multiprocessing.process import active_children as active_children, current_process as current_process
 
 # These are technically functions that return instances of these Queue classes.
-# Using them as annotations is deprecated. Either use imports from
-# multiprocessing.queues or the aliases defined below. See #4266 for discussion.
+# Using them as annotations is deprecated. Use imports from multiprocessing.queues instead.
+# See #4266 for discussion.
```

Comment:
I think this comment is just outdated? As far as I can see, all these are classes in the implementation, at least in Python 3.7+. Was this maybe different in the past? In that case, I don't think we need the type aliases below.

Reply:
They became functions at some point, but it was reverted.

Reply:
Somehow I missed that you asked about the implementation, not about the stub. These are functions (actually methods) in the implementation:

```
>>> from multiprocessing import Queue
>>> Queue
<bound method BaseContext.Queue of <multiprocessing.context.DefaultContext object at 0x7f66f26d73d0>>
```

There is also an undocumented class that is different from the public and documented `Queue`:

```
>>> from multiprocessing.queues import Queue
>>> Queue
<class 'multiprocessing.queues.Queue'>
```

There's similarly e.g. a method `multiprocessing.Event` and a class `multiprocessing.synchronize.Event`.

Reply:
This is what happens at runtime:

```pycon
>>> import multiprocessing
>>> multiprocessing.Queue
<bound method BaseContext.Queue of <multiprocessing.context.DefaultContext object at 0x000002F32AF7E5C0>>
>>> multiprocessing.Queue[int]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'method' object is not subscriptable
```

So the stubs are a bit of a lie at the moment. As @Akuli says, they were made into functions in #4289, but then that was reverted in #4314 after Jukka complained in #4313.

The thing is, either way, I don't think having these aliases helps the situation much. To work around these issues, a user can probably just put `multiprocessing.Queue` in quotes in their annotations, or use `from __future__ import annotations`.

Reply:
We should try to make them into functions again in the stubs and see the fallout, now that we have mypy_primer. Annotating something with `muliprocessing.Queue`, which is a function, is not something we should support.

Reply:
> We should try to make them into functions again in the stubs and see the fallout, now that we have mypy_primer. Annotating something with `muliprocessing.Queue`, which is a function, is not something we should support.

Sure. But not in this PR :)

Reply:
Should be a separate PR, but since the type aliases are connected to this, we should keep those that are alternatives to a function. We don't need to keep type aliases that are only an alternative for a class anyway.

Reply:
And going back to the controversy we had before: A typical user likely doesn't care that `multiprocessing.Queue` isn't a function. It's a technicality that gets in the way mostly when doing type annotations and when doing `isinstance(foo, Queue)`. My argument was that because `isinstance` checking is quite rare, including this technicality in the stubs is not a "practicality beats purity" solution. But I agree that we should see what mypy_primer says :)

For this PR, should we delete the aliases? We already know that no projects in mypy_primer are using them. I don't expect anyone else to be using them either, because annotating as if they were classes more or less works now.

Reply:
I marked my previous comment as off-topic, as it doesn't belong at all to this PR. I somehow didn't see other people's comments that were posted before I wrote mine.

Reply:
> since the type aliases are connected to this, we should keep those that are alternatives to a function.

I don't really see how the aliases help, even the ones which are alternatives to a function.

If it's a choice between doing this:

```python
from multiprocessing.synchronize import Lock
foo: Lock
```

or this:

```python
from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from multiprocessing import _LockType
foo: "_LockType"
```

Then, I find the first way much cleaner and more ergonomic. And, as @Akuli says in his minimized comment, we know none of the projects in mypy_primer are using these aliases.

Reply:
Sorry for being unclear. Yes, if there is a valid type somewhere, like `multiprocessing.synchronize.Lock`, we don't need the alias. It's possible that this is true for all aliases, I didn't check.

Reply:
`_BarrierType`, `_BoundedSemaphoreType`, `_ConditionType`, `_EventType`, `LockType`, `_RLockType` and `_SemaphoreType` are all aliases to valid types in `multiprocessing.synchronize`.

`_QueueType`, `_SimpleQueueType` and `_JoinableQueueType` are all aliases to valid types in `multiprocessing.queues`.

So yeah, it's true for all the aliases :)